lib/logstash/inputs/kafka.rb

```diff
@@ -133,31 +133,32 @@ def run(logstash_queue)
     @logger.info('Running kafka', :group_id => @group_id, :topic_id => @topic_id, :zk_connect => @zk_connect)
     begin
       @consumer_group.run(@consumer_threads,@kafka_client_queue)
-      begin
-        while true
-          event = @kafka_client_queue.pop
-          queue_event(event, logstash_queue)
-        end
-      rescue LogStash::ShutdownSignal
-        @logger.info('Kafka got shutdown signal')
-        @consumer_group.shutdown
+
+      while !stop?
+        event = @kafka_client_queue.pop
```

Comment:
What I used in other plugins is the non blocking version of pop, and then you can check if there is an exception (http://ruby-doc.org/stdlib-2.0.0/libdoc/thread/rdoc/Queue.html#method-i-pop) or I guess you could also check for empty? beforehand as you said.